mitmproxy/websocket.py

```diff
@@ -0,0 +1,51 @@
+import time
+
+from typing import List
+
+from mitmproxy import flow
+from mitmproxy.types import serializable
+
+
+class WebSocketMessage(serializable.Serializable):
+
+    def __init__(self, from_client, content, timestamp=None):
+        self.content = content
+        self.from_client = from_client
+        self.timestamp = timestamp or time.time()
+
+    @classmethod
+    def from_state(cls, state):
+        return cls(*state)
+
+    def get_state(self):
+        return self.from_client, self.content, self.timestamp
+
+    def set_state(self, state):
+        self.from_client = state.pop("from_client")
+        self.content = state.pop("content")
+        self.timestamp = state.pop("timestamp")
+
+    def __repr__(self):
+        return "{direction} {content}".format(
+            direction="->" if self.from_client else "<-",
+            content=repr(self.content)
+        )
+
+
+class WebSocketFlow(flow.Flow):
```

Comment:
The WebsocketFlow class should also contain a record of the handshake that initiated the websocket connection. That way, we can re-establish the connection if, for instance, we're replaying a websocket message in a websocket connection that has already been disconnected. The URL for the initial handshake is also the "end point" that will be shown in places like the flow list in the GUIs.